File: python/src/gibberish_rewriter/core/text_converter.py
"""Converts selected text to the other layout."""

from __future__ import annotations

import unicodedata
from dataclasses import dataclass

from .key_names import ascii_lower
from .key_table import KeyState, KeyTable, LayoutKind
from .word_list import WordList

_PLAIN_THEN_SHIFT = (KeyState.PLAIN, KeyState.SHIFT)
_MAX_SEQUENCES_CHECKED = 6
_WORD_BREAKS = " \t\r\n"


@dataclass(frozen=True)
class Conversion:
    """A converted selection, and the layout its text belongs to now."""

    text: str
    target: LayoutKind


@dataclass(frozen=True)
class _Piece:
    """A converted piece of a word: fixed text, or a two-character sequence one key or two keys could have typed."""

    one_key: str
    two_keys: str | None


def _is_latin_letter(c: str) -> bool:
    return "A" <= c <= "Z" or "a" <= c <= "z"


def _is_arabic_letter(c: str) -> bool:
    return "؀" <= c <= "ۿ" and unicodedata.category(c).startswith("L")


class TextConverter:
    """Converts selected text to the other layout."""

    def __init__(self, table: KeyTable, words: WordList) -> None:
        self._table = table
        self._words = words
        self._latin_typeable: set[str] = set()
        """Every character the Latin layout types."""
        self._arabic_texts: list[tuple[str, str]] = []
        """Every non-empty Arabic text with the Latin text of the same key and state, in table order, plain before shift."""
        for entry in table.keys:
            for text in entry.latin:
                self._latin_typeable.update(text)
            for state in _PLAIN_THEN_SHIFT:
                arabic = entry.text_on(LayoutKind.ARABIC, state)
                if arabic:
                    self._arabic_texts.append((arabic, entry.text_on(LayoutKind.LATIN, state)))

    def convert(self, text: str) -> Conversion | None:
        """Converts text, or returns None when it has no Latin or Arabic letters."""
        latin_letters = 0
        arabic_letters = 0
        last_letter: LayoutKind | None = None
        for c in text:
            if _is_latin_letter(c):
                latin_letters += 1
                last_letter = LayoutKind.LATIN
            elif _is_arabic_letter(c):
                arabic_letters += 1
                last_letter = LayoutKind.ARABIC
        if last_letter is None:
            return None
        if arabic_letters > latin_letters:
            source = LayoutKind.ARABIC
        elif latin_letters > arabic_letters:
            source = LayoutKind.LATIN
        else:
            source = last_letter
        if source is LayoutKind.LATIN:
            return Conversion(self._latin_to_arabic(text, latin_letters), LayoutKind.ARABIC)
        return Conversion(self._arabic_to_latin(text), LayoutKind.LATIN)

    def _latin_to_arabic(self, text: str, latin_letters: int) -> str:
        """Latin text to Arabic: the text was typed on the Latin layout while Arabic was meant."""
        all_caps = latin_letters >= 2 and not any("a" <= c <= "z" for c in text)
        result: list[str] = []
        for c in text:
            upper = "A" <= c <= "Z"
            found = self._table.find(ascii_lower(c) if upper else c, LayoutKind.LATIN)
            if found is None:
                result.append(c)
                continue
            entry, key_state = found
            state = key_state if not upper else KeyState.CAPS if all_caps else KeyState.SHIFT
            arabic = entry.text_on(LayoutKind.ARABIC, state)
            result.append(arabic if arabic else c)
        return "".join(result)

    def _arabic_to_latin(self, text: str) -> str:
        """Arabic text to Latin: the text was typed on the Arabic layout while Latin was meant."""
        result: list[str] = []
        word_start = -1
        for i in range(len(text) + 1):
            at_break = i == len(text) or text[i] in _WORD_BREAKS
            if not at_break:
                if word_start < 0:
                    word_start = i
                continue
            if word_start >= 0:
                result.append(self._convert_word(text[word_start:i]))
                word_start = -1
            if i < len(text):
                result.append(text[i])
        return "".join(result)

    def _convert_word(self, word: str) -> str:
        pieces: list[_Piece] = []
        i = 0
        while i < len(word):
            match = None if word[i] in self._latin_typeable else self._longest_arabic_match(word, i)
            if match is not None:
                arabic, latin = match
                one_key = latin if latin else arabic
                pieces.append(_Piece(one_key, self._two_keys(arabic) if len(arabic) == 2 else None))
                i += len(arabic)
            else:
                pieces.append(_Piece(word[i], None))
                i += 1

        sequences = sum(1 for piece in pieces if piece.two_keys is not None)
        if sequences == 0 or sequences > _MAX_SEQUENCES_CHECKED:
            return _build(pieces, 0, sequences)
        for combination in range(1 << sequences):
            candidate = _build(pieces, combination, sequences)
            if self._is_english_word(candidate):
                return candidate
        return _build(pieces, 0, sequences)

    def _two_keys(self, sequence: str) -> str | None:
        """The sequence read as two keys, or None when a character has no key of its own."""
        first = self._single_character_latin(sequence[0])
        second = self._single_character_latin(sequence[1])
        return None if first is None or second is None else first + second

    def _single_character_latin(self, c: str) -> str | None:
        for arabic, latin in self._arabic_texts:
            if arabic == c and latin:
                return latin
        return None

    def _is_english_word(self, candidate: str) -> bool:
        """ASCII-lowercases, strips leading and trailing characters other than a-z, and checks the word list."""
        lower = ascii_lower(candidate)
        start = 0
        end = len(lower)
        while start < end and not "a" <= lower[start] <= "z":
            start += 1
        while end > start and not "a" <= lower[end - 1] <= "z":
            end -= 1
        return end > start and self._words.contains(lower[start:end])

    def _longest_arabic_match(self, text: str, start: int) -> tuple[str, str] | None:
        """The longest Arabic-layout text starting at start; ties go to the earlier one."""
        best: tuple[str, str] | None = None
        for candidate in self._arabic_texts:
            if len(candidate[0]) > (len(best[0]) if best else 0) and text.startswith(candidate[0], start):
                best = candidate
        return best


def _build(pieces: list[_Piece], combination: int, sequences: int) -> str:
    """Bit 1 reads a sequence as two keys. The leftmost sequence is the most significant bit."""
    parts: list[str] = []
    bit = sequences - 1
    for piece in pieces:
        if piece.two_keys is None:
            parts.append(piece.one_key)
            continue
        parts.append(piece.two_keys if (combination >> bit) & 1 else piece.one_key)
        bit -= 1
    return "".join(parts)

File: csharp/src/GibberishRewriter.Core/TextConverter.cs
using System.Globalization;
using System.Text;

namespace GibberishRewriter.Core;

/// <summary>A converted selection, and the layout its text belongs to now.</summary>
public sealed record Conversion(string Text, LayoutKind Target);

/// <summary>Converts selected text to the other layout.</summary>
public sealed class TextConverter
{
    private static readonly KeyState[] PlainThenShift = [KeyState.Plain, KeyState.Shift];

    private const int MaxSequencesChecked = 6;

    private readonly WordList _words;

    private readonly KeyTable _table;

    /// <summary>Every character the Latin layout types.</summary>
    private readonly HashSet<char> _latinTypeable = [];

    /// <summary>Every non-empty Arabic text with the Latin text of the same key and state, in table order, plain before shift.</summary>
    private readonly List<(string Arabic, string Latin)> _arabicTexts = [];

    public TextConverter(KeyTable table, WordList words)
    {
        _table = table;
        _words = words;
        foreach (var entry in table.Keys)
        {
            foreach (var text in entry.Latin)
            {
                _latinTypeable.UnionWith(text);
            }
            foreach (var state in PlainThenShift)
            {
                var arabic = entry.TextOn(LayoutKind.Arabic, state);
                if (arabic.Length > 0)
                {
                    _arabicTexts.Add((arabic, entry.TextOn(LayoutKind.Latin, state)));
                }
            }
        }
    }

    /// <summary>Converts <paramref name="text"/>, or returns null when it has no Latin or Arabic letters.</summary>
    public Conversion? Convert(string text)
    {
        var latinLetters = 0;
        var arabicLetters = 0;
        LayoutKind? lastLetter = null;
        foreach (var c in text)
        {
            if (IsLatinLetter(c))
            {
                latinLetters++;
                lastLetter = LayoutKind.Latin;
            }
            else if (IsArabicLetter(c))
            {
                arabicLetters++;
                lastLetter = LayoutKind.Arabic;
            }
        }
        if (lastLetter is not { } last)
        {
            return null;
        }
        var source = arabicLetters > latinLetters ? LayoutKind.Arabic
            : latinLetters > arabicLetters ? LayoutKind.Latin
            : last;
        return source == LayoutKind.Latin
            ? new Conversion(LatinToArabic(text, latinLetters), LayoutKind.Arabic)
            : new Conversion(ArabicToLatin(text), LayoutKind.Latin);
    }

    private static bool IsLatinLetter(char c) => char.IsAsciiLetter(c);

    private static bool IsUpperLatin(char c) => char.IsAsciiLetterUpper(c);

    private static bool IsArabicLetter(char c) =>
        c is >= '\u0600' and <= '\u06FF'
        && CharUnicodeInfo.GetUnicodeCategory(c) is UnicodeCategory.UppercaseLetter
            or UnicodeCategory.LowercaseLetter
            or UnicodeCategory.TitlecaseLetter
            or UnicodeCategory.ModifierLetter
            or UnicodeCategory.OtherLetter;

    /// <summary>Latin text to Arabic: the text was typed on the Latin layout while Arabic was meant.</summary>
    private string LatinToArabic(string text, int latinLetters)
    {
        var allCaps = latinLetters >= 2 && !text.Any(char.IsAsciiLetterLower);
        var result = new StringBuilder(text.Length);
        foreach (var c in text)
        {
            var upper = IsUpperLatin(c);
            var lookup = upper ? char.ToLowerInvariant(c).ToString() : c.ToString();
            if (_table.Find(lookup, LayoutKind.Latin) is not { } key)
            {
                result.Append(c);
                continue;
            }
            var state = !upper ? key.State : allCaps ? KeyState.Caps : KeyState.Shift;
            var arabic = key.Entry.TextOn(LayoutKind.Arabic, state);
            if (arabic.Length > 0)
            {
                result.Append(arabic);
            }
            else
            {
                result.Append(c);
            }
        }
        return result.ToString();
    }

    /// <summary>A converted piece of a word: fixed text, or a two-character sequence one key or two keys could have typed.</summary>
    private readonly record struct Piece(string OneKey, string? TwoKeys);

    /// <summary>Arabic text to Latin: the text was typed on the Arabic layout while Latin was meant.</summary>
    private string ArabicToLatin(string text)
    {
        var result = new StringBuilder(text.Length);
        var wordStart = -1;
        for (var i = 0; i <= text.Length; i++)
        {
            var atBreak = i == text.Length || text[i] is ' ' or '\t' or '\r' or '\n';
            if (!atBreak)
            {
                if (wordStart < 0)
                {
                    wordStart = i;
                }
                continue;
            }
            if (wordStart >= 0)
            {
                result.Append(ConvertWord(text[wordStart..i]));
                wordStart = -1;
            }
            if (i < text.Length)
            {
                result.Append(text[i]);
            }
        }
        return result.ToString();
    }

    private string ConvertWord(string word)
    {
        var pieces = new List<Piece>();
        var i = 0;
        while (i < word.Length)
        {
            if (!_latinTypeable.Contains(word[i]) && LongestArabicMatch(word, i) is { } match)
            {
                var oneKey = match.Latin.Length > 0 ? match.Latin : match.Arabic;
                pieces.Add(new Piece(oneKey, match.Arabic.Length == 2 ? TwoKeys(match.Arabic) : null));
                i += match.Arabic.Length;
            }
            else
            {
                pieces.Add(new Piece(word[i].ToString(), null));
                i++;
            }
        }

        var sequences = pieces.Count(piece => piece.TwoKeys is not null);
        if (sequences is 0 or > MaxSequencesChecked)
        {
            return Build(pieces, 0, sequences);
        }
        for (var combination = 0; combination < 1 << sequences; combination++)
        {
            var candidate = Build(pieces, combination, sequences);
            if (IsEnglishWord(candidate))
            {
                return candidate;
            }
        }
        return Build(pieces, 0, sequences);
    }

    /// <summary>The sequence read as two keys, or null when a character has no key of its own.</summary>
    private string? TwoKeys(string sequence)
    {
        var first = SingleCharacterLatin(sequence[0]);
        var second = SingleCharacterLatin(sequence[1]);
        return first is null || second is null ? null : first + second;
    }

    private string? SingleCharacterLatin(char c)
    {
        foreach (var (arabic, latin) in _arabicTexts)
        {
            if (arabic.Length == 1 && arabic[0] == c && latin.Length > 0)
            {
                return latin;
            }
        }
        return null;
    }

    /// <summary>Bit 1 reads a sequence as two keys. The leftmost sequence is the most significant bit.</summary>
    private static string Build(List<Piece> pieces, int combination, int sequences)
    {
        var text = new StringBuilder();
        var bit = sequences - 1;
        foreach (var piece in pieces)
        {
            if (piece.TwoKeys is null)
            {
                text.Append(piece.OneKey);
                continue;
            }
            text.Append(((combination >> bit) & 1) == 1 ? piece.TwoKeys : piece.OneKey);
            bit--;
        }
        return text.ToString();
    }

    /// <summary>ASCII-lowercases, strips leading and trailing characters other than a–z, and checks the word list.</summary>
    private bool IsEnglishWord(string candidate)
    {
        var lower = string.Create(candidate.Length, candidate, static (span, source) =>
        {
            for (var i = 0; i < source.Length; i++)
            {
                span[i] = char.IsAsciiLetterUpper(source[i]) ? (char)(source[i] + 32) : source[i];
            }
        });
        var start = 0;
        var end = lower.Length;
        while (start < end && !char.IsAsciiLetterLower(lower[start]))
        {
            start++;
        }
        while (end > start && !char.IsAsciiLetterLower(lower[end - 1]))
        {
            end--;
        }
        return end > start && _words.Contains(lower[start..end]);
    }

    /// <summary>The longest Arabic-layout text starting at <paramref name="start"/>; ties go to the earlier one.</summary>
    private (string Arabic, string Latin)? LongestArabicMatch(string text, int start)
    {
        (string Arabic, string Latin)? best = null;
        foreach (var candidate in _arabicTexts)
        {
            if (candidate.Arabic.Length > (best?.Arabic.Length ?? 0)
                && text.AsSpan(start).StartsWith(candidate.Arabic, StringComparison.Ordinal))
            {
                best = candidate;
            }
        }
        return best;
    }
}
